packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -869,49 +856,55 @@ class HeroController extends NavigatorObserver {
   void _maybeStartHeroTransition(
     Route<dynamic>? fromRoute,
     Route<dynamic>? toRoute,
-    HeroFlightDirection flightType,
     bool isUserGestureTransition,
   ) {
     if (toRoute == fromRoute ||
         toRoute is! PageRoute<dynamic> ||
         fromRoute is! PageRoute<dynamic>) {
       return;
     }
-
-    final PageRoute<dynamic> from = fromRoute;
-    final PageRoute<dynamic> to = toRoute;
+    final Animation<double> newRouteAnimation = toRoute.animation!;
+    final Animation<double> oldRouteAnimation = fromRoute.animation!;
+    final HeroFlightDirection flightType;
+    if (isUserGestureTransition || oldRouteAnimation.status == AnimationStatus.reverse) {
+      flightType = HeroFlightDirection.pop;
+    } else if (newRouteAnimation.status == AnimationStatus.forward) {
+      flightType = HeroFlightDirection.push;
+    } else {
+      return;
```

Comment:
Use switch/case for enums: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-if-chains-or-or-with-enum-values